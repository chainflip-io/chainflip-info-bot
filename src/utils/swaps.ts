import { subMilliseconds } from 'date-fns';
import { chainConstants, type ChainflipChain } from './chainflip.js';
import { differenceInTimeAgo, intervalToDurationWords } from './date.js';

export const getSwapCompletionTime = ({
  depositChannelCreationTimestamp,
  preDepositBlockTimestamp,
  depositTimestamp,
  egressTimestamp,
  sourceChain,
  exact = true,
}: {
  depositTimestamp: Date;
  egressTimestamp: Date;
  sourceChain: ChainflipChain;
  depositChannelCreationTimestamp?: Date;
  preDepositBlockTimestamp?: Date;
  exact?: boolean;
}) => {
  let millisecondsToSubtract;
  if (preDepositBlockTimestamp) {
    const preDepositTimestamp = depositChannelCreationTimestamp
      ? new Date(
          Math.max(preDepositBlockTimestamp.getTime(), depositChannelCreationTimestamp.getTime()),
        )
      : preDepositBlockTimestamp;

    millisecondsToSubtract = (depositTimestamp.getTime() - preDepositTimestamp.getTime()) / 2;
  } else {
    millisecondsToSubtract = (chainConstants[sourceChain].blockTimeSeconds / 2) * 1000;
  }

  return exact
    ? intervalToDurationWords({
        start: subMilliseconds(depositTimestamp, millisecondsToSubtract).getTime(),
        end: egressTimestamp.getTime(),
      })
    : differenceInTimeAgo(
        subMilliseconds(depositTimestamp, millisecondsToSubtract).toISOString(),
        false,
        egressTimestamp.toISOString(),
      );
};
