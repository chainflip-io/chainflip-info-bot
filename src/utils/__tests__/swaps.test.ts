/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { describe, expect, it } from 'vitest';
import { getSwapCompletionTime } from '../swaps';

describe('getSwapCompletionTime', () => {
  describe('swap completion time estimate based on statically defined pre-block time intervals', () => {
    it.each([
      ['Ethereum', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Bitcoin', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Arbitrum', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Polkadot', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
    ])(
      `returns an exact swap completion time estimate for %s`,
      (sourceChain, depositTimestamp, egressTimestamp) => {
        const completionTime = getSwapCompletionTime({
          depositTimestamp,
          egressTimestamp,
          sourceChain,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
    it.each([
      ['Ethereum', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Bitcoin', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Arbitrum', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
      ['Polkadot', new Date('2022-01-01T00:00:00Z'), new Date('2022-01-01T00:05:00Z')] as const,
    ])(
      `returns a truncated swap completion time estimate for %s`,
      (sourceChain, depositTimestamp, egressTimestamp) => {
        const completionTime = getSwapCompletionTime({
          depositTimestamp,
          egressTimestamp,
          sourceChain,
          exact: false,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
  });

  describe('swap completion time estimate based on dynamically defined pre-block time intervals', () => {
    it.each([
      [
        'Ethereum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns an exact swap completion time estimate for %s`,
      (sourceChain, preDepositBlockTimestamp, depositTimestamp, egressTimestamp) => {
        const completionTime = getSwapCompletionTime({
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
    it.each([
      [
        'Ethereum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns a truncated swap completion time estimate for %s`,
      (sourceChain, preDepositBlockTimestamp, depositTimestamp, egressTimestamp) => {
        const completionTime = getSwapCompletionTime({
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
          exact: false,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
  });
  describe('swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp', () => {
    it.each([
      [
        'Ethereum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns an exact swap completion time estimate for %s`,
      (
        sourceChain,
        preDepositBlockTimestamp,
        depositChannelCreationTimestamp,
        depositTimestamp,
        egressTimestamp,
      ) => {
        const completionTime = getSwapCompletionTime({
          depositChannelCreationTimestamp,
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
    it.each([
      [
        'Ethereum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:30Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns a truncated swap completion time estimate for %s`,
      (
        sourceChain,
        preDepositBlockTimestamp,
        depositChannelCreationTimestamp,
        depositTimestamp,
        egressTimestamp,
      ) => {
        const completionTime = getSwapCompletionTime({
          depositChannelCreationTimestamp,
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
          exact: false,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
  });
  describe('swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp', () => {
    it.each([
      [
        'Ethereum',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns an exact swap completion time estimate for %s`,
      (
        sourceChain,
        depositChannelCreationTimestamp,
        preDepositBlockTimestamp,
        depositTimestamp,
        egressTimestamp,
      ) => {
        const completionTime = getSwapCompletionTime({
          depositChannelCreationTimestamp,
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
    it.each([
      [
        'Ethereum',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Bitcoin',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Arbitrum',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
      [
        'Polkadot',
        new Date('2021-01-01T00:00:00Z'),
        new Date('2022-01-01T00:00:00Z'),
        new Date('2022-01-01T00:01:00Z'),
        new Date('2022-01-01T00:05:00Z'),
      ] as const,
    ])(
      `returns a truncated swap completion time estimate for %s`,
      (
        sourceChain,
        depositChannelCreationTimestamp,
        preDepositBlockTimestamp,
        depositTimestamp,
        egressTimestamp,
      ) => {
        const completionTime = getSwapCompletionTime({
          depositChannelCreationTimestamp,
          preDepositBlockTimestamp,
          depositTimestamp,
          egressTimestamp,
          sourceChain,
          exact: false,
        });

        expect(completionTime).toMatchSnapshot();
      },
    );
  });
});
