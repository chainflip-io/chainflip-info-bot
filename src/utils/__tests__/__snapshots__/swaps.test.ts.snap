// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns a truncated swap completion time estimate for Arbitrum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns a truncated swap completion time estimate for Bitcoin 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns a truncated swap completion time estimate for Ethereum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns a truncated swap completion time estimate for Polkadot 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns an exact swap completion time estimate for Arbitrum 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns an exact swap completion time estimate for Bitcoin 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns an exact swap completion time estimate for Ethereum 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals > returns an exact swap completion time estimate for Polkadot 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns a truncated swap completion time estimate for Arbitrum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns a truncated swap completion time estimate for Bitcoin 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns a truncated swap completion time estimate for Ethereum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns a truncated swap completion time estimate for Polkadot 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns an exact swap completion time estimate for Arbitrum 1`] = `"04min 15s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns an exact swap completion time estimate for Bitcoin 1`] = `"04min 15s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns an exact swap completion time estimate for Ethereum 1`] = `"04min 15s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using channel creation timestamp > returns an exact swap completion time estimate for Polkadot 1`] = `"04min 15s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns a truncated swap completion time estimate for Arbitrum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns a truncated swap completion time estimate for Bitcoin 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns a truncated swap completion time estimate for Ethereum 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns a truncated swap completion time estimate for Polkadot 1`] = `"4 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns an exact swap completion time estimate for Arbitrum 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns an exact swap completion time estimate for Bitcoin 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns an exact swap completion time estimate for Ethereum 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on dynamically defined pre-block time intervals using pre deposit block timestamp > returns an exact swap completion time estimate for Polkadot 1`] = `"04min 30s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns a truncated swap completion time estimate for Arbitrum 1`] = `"5 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns a truncated swap completion time estimate for Bitcoin 1`] = `"10 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns a truncated swap completion time estimate for Ethereum 1`] = `"5 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns a truncated swap completion time estimate for Polkadot 1`] = `"5 min"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns an exact swap completion time estimate for Arbitrum 1`] = `"05min 00s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns an exact swap completion time estimate for Bitcoin 1`] = `"10min 00s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns an exact swap completion time estimate for Ethereum 1`] = `"05min 06s"`;

exports[`getSwapCompletionTime > swap completion time estimate based on statically defined pre-block time intervals > returns an exact swap completion time estimate for Polkadot 1`] = `"05min 03s"`;
